File: go/problem21.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(FindAmicableNumberSum(1000))
}

func FindAmicableNumberSum(limit int) int {

	factorSumMapSlice := make(map[int]int)
	for i := 2; i <= limit; i++ {
		factorSumMapSlice[i] = factorSum(i)
	}

	amicableNumberSum := 0
	for index, val := range factorSumMapSlice {
		if index == factorSumMapSlice[val] && index != val {
			// fmt.Println("Found an amicable number ", val, index, factorSumMapSlice[val])
			amicableNumberSum = amicableNumberSum + val + index
			factorSumMapSlice[val] = -1 // to avoid duplicate count or need to divide by 2
		}
	}

	return amicableNumberSum
}

func factorSum(number int) int {

	var factorSum int = 1
	var i int
	var incrementor int = 1
	if number%2 != 0 {
		incrementor = 2
	} else {
		quotient := number / 2
		if quotient != 2 {
			factorSum += quotient
		}
		factorSum += 2
	}

	for i = 3; i <= int(math.Sqrt(float64(number))); i += incrementor {
		if number%i == 0 {
			quotient := number / i
			if quotient != i {
				factorSum += quotient
			}
			factorSum += i
		}
	}

	return factorSum
}

File: python/problem21.py
import math

def findAmicableNumberSum(limit):
    factorSumMap = {}
    for i in range(2, limit + 1):
        factorSumMap[i] = factorSum(i)

    amicableNumberSum = 0
    for key in factorSumMap:
        val = factorSumMap.get(key)
        if val and key == factorSumMap.get(val) and key != val:
            amicableNumberSum = amicableNumberSum + val + key
            factorSumMap[val] = -1  # to avoid duplicate count or need to divide by 2

    return amicableNumberSum


def factorSum(number):
    factorSum = 1
    incrementor = 1
    if number % 2 != 0:
        incrementor = 2
    else:
        quotient = number / 2
        if quotient != 2:
            factorSum += quotient

        factorSum += 2

    for i in range(3, int(math.sqrt(number)), incrementor):
        if number % i == 0:
            quotient = number // i
            if quotient != i:
                factorSum += quotient

            factorSum += i

    return factorSum


if __name__ == '__main__':
    print(findAmicableNumberSum(1000))
